subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -0,0 +1,160 @@
+        setup_original_apple_certificate(settings.APPLE_ROOT_CERTIFICATE_PATH)
+
+    def charge_online(self, user: AbstractBaseUser, plan: Plan, subscription: Optional[Subscription] = None,
+                      quantity: int = 1) -> Tuple[SubscriptionPayment, str]:
+        """
+        In case of in-app purchase this operation is triggered from the mobile application library.
+        """
+        raise InvalidOperation()
+
+    def charge_offline(self, user: AbstractBaseUser, plan: Plan, subscription: Optional[Subscription] = None,
+                       quantity: int = 1,
+                       reference_payment: Optional[SubscriptionPayment] = None) -> SubscriptionPayment:
+        raise InvalidOperation()
+
+    def webhook(self, request: Request, payload: dict) -> Response:
+        if self.transaction_receipt_tag in payload:
+            return self._handle_receipt(request, payload)
+        elif self.signed_payload_tag in payload:
+            return self._handle_app_store(request, payload)
+        else:
+            # Invalid, unhandled request.
+            return Response(status=400)
+
+    def check_payments(self, payments: Iterable[SubscriptionPayment]):
+        pass
+
+    @staticmethod
+    def _get_validated_in_app_product(response: AppleVerifyReceiptResponse) -> AppleInApp:
+        assert response.is_valid, str(response)
+        assert response.receipt.bundle_id == settings.APPLE_BUNDLE_ID, str(response)
+        assert len(response.receipt.in_apps) == 1
+        return response.receipt.in_apps[0]
+
+    def _handle_receipt(self, request: Request, payload: dict) -> Response:
+        receipt = payload[self.transaction_receipt_tag]
+
+        # Validate the receipt. Fetch the status and product.
+        receipt_data = self.api.fetch_receipt_data(receipt)
+        single_in_app = self._get_validated_in_app_product(receipt_data)
+
+        # Check whether this receipt is anyhow interesting:
+        try:
+            payment = SubscriptionPayment.objects.get(provider_codename=self.codename,
+                                                      provider_transaction_id=single_in_app.transaction_id)
+            # We've already handled this. And all the messages coming to us from iOS should be AFTER
+            # the client money were removed.
+            # TODO(kkalinowski): return the subscription payment object.
+            return Response()
+        except SubscriptionPayment.DoesNotExist:
+            # If it doesn't exist, it's all right – most probably it needs to be created.
+            pass
+
+        # Find the right plan to create subscription.
+        plan = Plan.objects.get(apple_in_app=single_in_app.product_id)
+
+        # Create subscription.
+        subscription = Subscription(
+            user=request.user,
+            plan=plan,
+            # For in-app purchases this option doesn't make sense.
+            auto_prolong=False,
+        )
+        subscription.save()
```

Comment:
Why using `objects.create` in this case? Won't it perform `insert` to the database, then we modify the object, then we perform another `update` on the database? I know it'll all be wrapped in a transaction, so it doesn't matter, but wouldn't the normal instantiation be "better" in this case? (trying to understand Django a little bit better)

Reply:
If you can do something like
```python
instance = Model(...)
instance.save()
```
then you can do the same with
```
instance = Model.objects.create(...)
```
It's all the same.

In this particular case:
1) You create a Payment (no matter whether you use `save()` or `create()`, it's 1 query)
2) the payment creates a Subscription automatically on save (1 query) - this is essential part because payment adjusts subscription duration according to its own duration, plus it won't create subscription if its status is not COMPLETED - so it does some nice logic for handling Subscription creation and you have to use it
3) you modify the Subscription (disable auto-prolong) - (1 query)

You could do 1) create a Subscription manually, and then 2) create a Payment and link it to manually created subscription, and it will be 2 queries in total, but you lose that payment-auto-save logic.

Ping me on slack if you have more questions on this

Reply:
Done.